### PR TITLE
feat: support `neq` and `fieldref|neq` field modifiers (#1684)

### LINKED ISSUE #1684: Support `neq` and `
In the new 2.1 specification https://github.com/SigmaHQ/sigma-specification/blob/main/specification/sigma-appendix-modifiers.md , the `neq` modifier has been added.

`neq`: The field is different from the specified values.

It can also be used with `fieldref` (`|fieldref|neq:`) to see if two field values are different.
`fieldref: Modifies a plain string into a field reference. A field reference can be used to compare fields of matched events directly at query/matching time. Can be conbine with the neq modifier.`

Since we already use this logic in our rules, I would like to support this so we can update our rules to the official format.

@fukusuket Could you take a look at this when you get a chance?

### PR DESCRIPTION
**Pull request by Shirofune-Security**
## What Changed

Adds support for the Sigma 2.1 [`neq`](https://github.com/SigmaHQ/sigma-specification/blob/main/specification/sigma-appendix-modifiers.md) modifier and its combination with `fieldref`, as requested in #1684.

`neq` negates the whole comparison — mirroring pySigma's `SigmaNegateModifier` (which sets `detection_item.negated = True`) rather than being a value transform. Because of this it composes with **any** other modifier:

- `Field|neq: value` — field is different from the value
- `Field|fieldref|neq: OtherField` — the two fields have different values
- and freely with the rest: `|contains|neq`, `|startswith|neq`, `|endswith|neq`, `|re|neq`, `|re|i|neq`, `|cidr|neq`, `|gt|neq` (and `lt`/`gte`/`lte`), `|contains|cased|neq`, `|fieldref|contains|neq`, …

### Implementation

- **`matchers.rs`** — `DefaultMatcher` gains a `neg_match` flag. In `init()` the `neq` token is detected among the pipe modifiers (never the field name) and stripped from the chain, setting the flag; the rest of parsing is completely unchanged. `is_match()` now returns `is_match_inner(...) ^ neg_match`, so negation is applied once, over the fully-resolved result of every match path (fast-match, regex, fieldref, numeric, cidr, …).
- **`mod.rs`** — a value **list** under a `neq` key is combined with the De Morgan-correct connective: a normal (OR-linked) list becomes AND-of-negations (`NOT(a OR b) = ¬a AND ¬b`), and an `|all` (AND-linked) list becomes OR-of-negations (`NOT(a AND b) = ¬a OR ¬b`).
- **`selectionnodes.rs`** — for multi-valued `EventData.Data` arrays the negation is applied once over the whole field (`all` instead of `any`, via a new `LeafMatcher::is_negated()`), so a rule and its literal negation can never both match the same record.

A field that does not exist is treated as *different* from the value (so `neq` matches), consistent with how condition-level `not` already behaves in Hayabusa.

Docs (`field modifiers` page) and the CHANGELOG are updated.

## Evidence

The change adds **41 unit tests** (all passing) covering each modifier combination and the edge cases:

- core behavior: `|neq` detect / not-detect, case-insensitivity, wildcards, missing field
- composed modifiers: `contains|neq`, `startswith|neq`, `endswith|neq`, `re|neq`, `re|i|neq`, `gt|neq` (equal / greater / missing), `cidr|neq`, `contains|cased|neq`
- field references: `fieldref|neq`, `fieldref|contains|neq`, missing referenced field
- value lists (De Morgan): OR-linked `|neq` list, AND-linked `|contains|all|neq` list
- multi-valued `EventData.Data` arrays with `|neq`
- repeated `neq` is idempotent (matches Sigma semantics)

```
$ cargo test --lib
test result: ok. 458 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
$ cargo fmt -- --check      # clean
$ cargo clippy --all-targets -- -D warnings   # clean
```

Closes #1684.

> Note: field-modifier documentation was updated in English (`website/docs/rules/detection-fields.md`); the localized translations can follow.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,8 @@
 
 **Enchancements:**
 
-Added unique and total alert count to MITRE ATT&CK tactics found. (#1753) (@fukusuket)
+- Added support for the `|neq` and `|fieldref|neq` field modifiers from the Sigma 2.1 specification. `|neq` negates a comparison and can be combined with other modifiers (`|contains|neq`, `|startswith|neq`, `|endswith|neq`, `|fieldref|neq`, ...). (#1684) (@Shirofune-Security)
+- Added unique and total alert count to MITRE ATT&CK tactics found. (#1753) (@fukusuket)
```
[the remark this answers is not in the document]
Comment (Shirofune-Security) [reply CHANGELOG.md]: Fixed in 52319ceb — corrected to "Enhancements" (matching the spelling used in the rest of the CHANGELOG).

File: website/docs/rules/detection-fields.md [hunk 2]
```diff
@@ -259,11 +259,13 @@ This document is dynamically updated every time there is an update to Sigma or H
...
 - `|fieldref|startswith`: Check if the field on the left starts with the string of the field on the right. You can use `not` in the `condition` to check if they are different.
 - `|gt`: Checks if a field value is greater than a certain number.
 - `|gte`: Checks if a field value is greater than or equal to a certain number.
 - `|lt`: Checks if a field value is less than a certain number.
 - `|lte`: Checks if a field value is less than or equal to a certain number.
+- `|neq`: Checks if a field value is different from the specified value (the negation of a plain value match). It can be combined with other modifiers (`|contains|neq`, `|startswith|neq`, `|endswith|neq`, `|fieldref|neq`, ...) to negate them. When multiple values are specified in a list, the field must be different from **all** of them. A field that does not exist is treated as different from the value and therefore matches.
 - `|re`: Use case-sensitive regular expressions. (We are using the regex crate so please out the documentation at <https://docs.rs/regex/latest/regex/#syntax> to learn how to write supported regular expressions.)
```
[the remark this answers is not in the document]
Comment (Shirofune-Security) [reply website/docs/rules/detection-fields.md]: Fixed in 52319ceb — changed to "please check out the documentation".

**Comment by YamatoSecurity** [comment]
@fukusuket Whenever you get time, could you check this?

**Review by fukusuket — COMMENTED** [review]
@YamatoSecurity 
I'm not sure if it's due to the diff in this PR, but I noticed a difference in the detection results between 3.9.0 and this PR. Could you check this?🙏

```
>./hayabusa csv-timeline -p super-verbose -s -w -q -C -o timeline-new.csv -d ../data/windows/hayabusa-sample-evtx-main
> ./hayabusa-3.9.0-mac-aarch64 csv-timeline -p super-verbose -s -w -q -C -o timeline-old.csv -d ../data/windows/hayabusa-sample-evtx-main
> ls -la
-rw-r--r--@  1 fukusuke  staff  40777377  7  3 07:32 timeline-new.csv
-rw-r--r--@  1 fukusuke  staff  40770941  7  3 07:33 timeline-old.csv
```
The following records only appear in 3.9.0.
```
< "2021-04-23 19:09:46.214 +09:00","Uncommon New Firewall Rule Added In Windows Firewall Exception List","med","srvdefender01.offsec.lan","Firewall",2004,"frack113","2025-10-08","test",434,"RuleName: OpenSSH Server (sshd) ¦ App: ¦ ModifyingApp: C:\Windows\System32\wbem\WmiPrvSE.exe ¦ User: S-1-5-21-4230534742-2542757381-3142984815-1111 ¦ Svc: ¦ Protocol: 6 ¦ LocalAddr: * ¦ LocalPort: 22 ¦ RemoteAddr: * ¦ RemotePort: * ¦ Action: 3","Active: 1 ¦ Channel: Microsoft-Windows-Windows Firewall With Advanced Security/Firewall ¦ Culture: en-US ¦ Direction: 1 ¦ EdgeTraversal: 0 ¦ Flags: 1 ¦ Level: Information ¦ LocalOnlyMapped: 0 ¦ LooseSourceMapped: 0 ¦ Message: A rule has been added to the Windows Firewall exception list. Added Rule: Rule ID: sshd Rule Name: OpenSSH Server (sshd) Origin: Local Active: Yes Direction: Inbound Profiles: Private,Domain, Public Action: Allow Application Path: Service Name: Protocol: TCP Security Options: None Edge Traversal: None Modifying User: S-1-5-21-4230534742-2542757381-3142984815-1111 Modifying Application: C:\Windows\System32\wbem\WmiPrvSE.exe ¦ Opcode: Info ¦ Origin: 1 ¦ Profiles: 2147483647 ¦ Provider: Microsoft-Windows-Windows Firewall With Advanced Security ¦ RuleId: sshd ¦ RuleStatus: 65536 ¦ SchemaVersion: 538 ¦ SecurityOptions: 0","DefImpair","T1686.003","","Firewall","2022-02-19","win_firewall_as_add_rule.yml","a1dbd390-a4c5-755a-1b1f-a76aabbecbbc","../data/windows/hayabusa-sample-evtx-main/EVTX-to-MITRE-Attack/TA0005-Defense Evasion/T1562.004-Impair Defenses-Disable or Modify System Firewall/ID4103-4104-2004-OpenSSH firewall rule activation.evtx"
```
**Uncommon New Firewall Rule Added In Windows Firewall Exception List**
- https://github.com/Yamato-Security/hayabusa-rules/blob/1135904dd76c8eef7c7cf7b1d6ddc20509793462/sigma/builtin/firewall_as/win_firewall_as_add_rule.yml#L53

**Comment by YamatoSecurity** [comment]
@fukusuket Yes, the new rule was due to a bug in the evtx crate. I fixed it in https://github.com/Yamato-Security/hayabusa-evtx/pull/90
The main branch of hayabusa should now also detect this so when you are doing comparisons, please use the `main` branch to compare against instead of 3.9.0 because there have been quite a few changes.

**Comment by fukusuket** [comment]
@YamatoSecurity 
Regarding https://github.com/Yamato-Security/hayabusa-evtx/pull/90 , I might be misunderstanding, but I actually thought the behavior before the fix was the expected one. Could you clarify?I had assumed that we should use the [exists modifier](https://sigmahq.io/docs/basics/modifiers.html#exists) to check for the presence of a field.

Separately from that topic, I will compare this PR's modifications with the main branch :)

**Comment by YamatoSecurity** [comment]
> @YamatoSecurity Regarding [Yamato-Security/hayabusa-evtx#90](https://github.com/Yamato-Security/hayabusa-evtx/pull/90) , I might be misunderstanding, but I actually thought the behavior before the fix was the expected one. Could you clarify?I had assumed that we should use the [exists modifier](https://sigmahq.io/docs/basics/modifiers.html#exists) to check for the presence of a field.
> 
> Separately from that topic, I will compare this PR's modifications with the main branch :)

I think the `|exists` field modifier is relatively new and is meant to replace the use of `null`.
I was a little confused about this at first and thought that the older evtx crate was the correct behavior but I think it isn't now.
Here are the ways to create empty and null fields.
Screenshot 2026-07-04 at 11 53 22

The first one is the explicit way but Windows event logs do not do this.

`</Field>` and `<Field></Field>` have a field named `Field` but the value is an empty string. The field still does exist in the log. It is the same as `Field|exists: true` but I believe you can also do the same logic with:
```
   filter:
       Field: null
condition: not filter
```

When `Field` does not exist in the log at all you can either specify this with `Field|exists: false` OR 
```
    filter:
         Field: null
condition: filter
```

I don't think there exists Windows event logs that do things like `<Field>""</Field>` or `<Field>''</Field>` and even if they did exist, they would be the same meaning as `<Field></Field>`, no?

Please let me know if you have seen these types of fields though. I am not 100% about all of this so let me know if you see some evidence that says otherwise.

**Comment by fukusuket** [comment]
@YamatoSecurity  
Thank you! Sorry, I was a bit concerned because an empty string (**is still a value**) is different from null (**the absence of a value**). However, I understand that we are operating under the assumption that such data will not occur in the Windows EventLog. Also, after re-checking the rule logic, it turns out that the original logs are actually NOT supposed to be detected.

And the  `neq` modifier works!! LGTM!!🚀

**Comment by YamatoSecurity** [comment]
@fukusuket Thanks so much for double-checking! Yes, it is confusing because an empty string and null are traditionally different in programming but I think with sigma `null` means the absence of the **field**, not absence of the **value**.

File: src/detections/rule/matchers.rs [hunk 3]
```diff
@@ -276,6 +287,24 @@ impl LeafMatcher for DefaultMatcher {
...
+            let field = keys_all[0];
+            let mut rest: Vec<&str> = keys_all[1..]
+                .iter()
+                .copied()
+                .filter(|k| *k != "neq")
+                .collect();
+            keys_all = Vec::with_capacity(rest.len() + 1);
+            keys_all.push(field);
+            keys_all.append(&mut rest);
+        }
```
[the remark this answers is not in the document]
Comment (Shirofune-Security) [reply src/detections/rule/matchers.rs:306]: Good catch — this was a real footgun. Fixed in d964ed09: `init()` now rejects `neq` on a keyless `|all` selection (empty field name) with a clear error, so `|all|neq` fails to load instead of falling through to an empty-field match and matching every record. Added a regression test (`test_neq_on_keyless_all_is_rejected`).

File: website/docs/rules/detection-fields.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shirofune-Security) [reply website/docs/rules/detection-fields.md]: Fixed — the text now reads "please check out the documentation" (in the docs-typo commit).

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shirofune-Security) [reply CHANGELOG.md]: `Enchancements` is a pre-existing typo in the already-released `3.9.0` section, not part of this PR's changes — this PR's own entry is under the correctly-spelled `Enhancements` header. I've left the released section untouched to keep this feature PR focused; happy to fix it in a separate docs PR if you'd like.

**Review by YamatoSecurity — APPROVED** [review]
LGTM
